File: packages/cli/src/format.ts
import type { ResolveResult } from '@vscode-links/core'

export type OutputFormat = 'json' | 'ndjson' | 'tsv'

/**
 * Render a ResolveResult to a string in one of the supported formats.
 *
 * - json:   pretty-printed object, one shot.
 * - ndjson: one JSON object per line, prefixed by kind:
 *           {"kind":"context",...}
 *           {"kind":"link","data":{...}}
 *           {"kind":"skipped","data":{...}}
 *           {"kind":"diagnostic","data":{...}}
 *   Easier for downstream pickers to ingest with minimal buffering.
 * - tsv:    tab-separated link rows for shell pipelines:
 *           type<TAB>source<TAB>url<TAB>title
 *           Skipped/diagnostic rows are NOT included; use ndjson for those.
 */
export function format(result: ResolveResult, mode: OutputFormat): string {
  switch (mode) {
    case 'json':
      return `${JSON.stringify(result, null, 2)}\n`
    case 'ndjson':
      return formatNdjson(result)
    case 'tsv':
      return formatTsv(result)
    default: {
      const exhaustive: never = mode
      throw new Error(`unknown format: ${exhaustive as string}`)
    }
  }
}

function formatNdjson(result: ResolveResult): string {
  const lines: string[] = []
  lines.push(JSON.stringify({ kind: 'context', data: result.context }))
  for (const link of result.links) {
    lines.push(JSON.stringify({ kind: 'link', data: link }))
  }
  for (const skipped of result.skipped) {
    lines.push(JSON.stringify({ kind: 'skipped', data: skipped }))
  }
  for (const diag of result.diagnostics) {
    lines.push(JSON.stringify({ kind: 'diagnostic', data: diag }))
  }
  return `${lines.join('\n')}\n`
}

function formatTsv(result: ResolveResult): string {
  const rows: string[] = []
  for (const link of result.links) {
    rows.push([
      link.type,
      link.source,
      link.url,
      link.title,
    ].map(escapeTsvField).join('\t'))
  }
  return rows.length === 0 ? '' : `${rows.join('\n')}\n`
}

// TSV cells must not contain tab or newline; replace defensively.
function escapeTsvField(s: string): string {
  return s.replace(/\t/g, ' ').replace(/\r?\n/g, ' ')
}

File: crates/cli/src/format.rs
use std::fmt::Write;

use serde::Serialize;
use vscode_links_core::ResolveResult;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OutputFormat {
    Json,
    Ndjson,
    Tsv,
}

impl std::str::FromStr for OutputFormat {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "json" => Ok(Self::Json),
            "ndjson" => Ok(Self::Ndjson),
            "tsv" => Ok(Self::Tsv),
            other => Err(format!("--format must be one of json|ndjson|tsv, got {other}")),
        }
    }
}

pub fn format_result(result: &ResolveResult, mode: OutputFormat) -> String {
    match mode {
        OutputFormat::Json => format!(
            "{}\n",
            serde_json::to_string_pretty(result).expect("serde infallible for ResolveResult")
        ),
        OutputFormat::Ndjson => format_ndjson(result),
        OutputFormat::Tsv => format_tsv(result),
    }
}

#[derive(Serialize)]
struct NdjsonEnvelope<'a, T: Serialize> {
    kind: &'static str,
    data: &'a T,
}

#[derive(Serialize)]
struct NdjsonContext<'a> {
    kind: &'static str,
    #[serde(flatten)]
    data: &'a serde_json::Value,
}

fn format_ndjson(result: &ResolveResult) -> String {
    let mut out = String::new();
    // context goes first; JS side uses { kind: "context", data: ... } too.
    let ctx = NdjsonEnvelope { kind: "context", data: &result.context };
    writeln!(out, "{}", serde_json::to_string(&ctx).unwrap()).unwrap();
    for link in &result.links {
        let envelope = NdjsonEnvelope { kind: "link", data: link };
        writeln!(out, "{}", serde_json::to_string(&envelope).unwrap()).unwrap();
    }
    for skipped in &result.skipped {
        let envelope = NdjsonEnvelope { kind: "skipped", data: skipped };
        writeln!(out, "{}", serde_json::to_string(&envelope).unwrap()).unwrap();
    }
    for diag in &result.diagnostics {
        let envelope = NdjsonEnvelope { kind: "diagnostic", data: diag };
        writeln!(out, "{}", serde_json::to_string(&envelope).unwrap()).unwrap();
    }
    let _ = NdjsonContext { kind: "context", data: &result.context }; // keep helper compiling
    out
}

fn format_tsv(result: &ResolveResult) -> String {
    let mut out = String::new();
    for link in &result.links {
        let type_str = serde_json::to_string(&link.r#type).unwrap();
        let type_unquoted = type_str.trim_matches('"');
        writeln!(
            out,
            "{}\t{}\t{}\t{}",
            escape_tsv(type_unquoted),
            escape_tsv(&link.source),
            escape_tsv(&link.url),
            escape_tsv(&link.title),
        )
        .unwrap();
    }
    out
}

fn escape_tsv(s: &str) -> String {
    s.replace('\t', " ").replace(['\n', '\r'], " ")
}
